apps/sms/test/marionette/mocks/mock_navigator_moz_mobile_message.js

```diff
@@ -27,215 +24,210 @@ Services.obs.addObserver(function(document) {
-        }, delayMs);
+      var requests = recipients.map((receiver) => {
+        var request = Services.DOMRequest.createRequest(window);
+
+        var parameters = {
+          type: 'sms',
+          body: content,
+          receiver: receiver
+        };
+
+        createMessage(parameters, options).then(function(message) {
+          Services.DOMRequest.fireSuccessAsync(request, null);
+
+          callEventHandlers(
+            'sending', Cu.cloneInto({ message: message }, window)
+          );
+        });
       });
 
-      return requests;
+      return Cu.waiveXrays(requests);
     },
 
     getThreads: function() {
-      var cursor = Cu.waiveXrays(Cu.cloneInto({
-        onsuccess: null,
-        onerror: null
-      }, window));
+      var threads = null;
+
+      var handleCursor = function() {
+        var iteratorResult = threads.next();
 
-      window.setTimeout(function() {
-        if (typeof cursor.onsuccess === 'function') {
-          cursor.onsuccess.call(cursor);
+        if (!iteratorResult.done) {
+          Services.DOMRequest.fireSuccessAsync(cursor, iteratorResult.value);
+        } else {
+          Services.DOMRequest.fireDone(cursor);
         }
-      }, delayMs);
+      };
+
+      var cursor = Services.DOMRequest.createCursor(window, handleCursor);
+
+      getStorage().then(function(storage) {
+        threads = storage.threads.values();
+
+        handleCursor();
+      });
 
       return cursor;
     },
 
     getMessages: function(filter) {
-      var cursor = Cu.waiveXrays(Cu.cloneInto({
-        onsuccess: null,
-        onerror: null,
-        result: null,
-        continue: null,
-        done: false
-      }, window));
+      var messages = null;
 
-      window.setTimeout(function() {
-        if (typeof cursor.onsuccess === 'function') {
-          var thread = threads.get(filter.threadId),
-              currentIndex = -1;
-
-          cursor.continue = function() {
-            if (!thread || ++currentIndex >= thread.messages.length) {
-              cursor.done = true;
-              cursor.continue = null;
-              cursor.result = null;
-            } else {
-              cursor.result = thread.messages[currentIndex];
-            }
-            cursor.onsuccess.call(cursor);
-          };
+      var handleCursor = function() {
+        var iteratorResult = messages.next();
 
-          cursor.continue();
+        if (!iteratorResult.done) {
+          Services.DOMRequest.fireSuccessAsync(cursor, iteratorResult.value);
+        } else {
+          Services.DOMRequest.fireDone(cursor);
         }
-      }, delayMs);
+      };
+
+      var cursor = Services.DOMRequest.createCursor(window, handleCursor);
+
+      getStorage().then(function(storage) {
+        var thread = storage.threads.get(filter.threadId);
+
+        // Remove this once Array.prototype.values() is landed (see bug 875433).
+        messages = new Set(thread && thread.messages || []).values();
```

Comment:
`(thread && thread.messages || [])[Symbol.iterator]()` ? Not sure this is really better :)

`(thread && thread.messages || []).entries()` is maybe better though. In that case you'll need `iteratorResult.value[1]`. Or: `var { done, value: [_, value] } = messages.next();` ;)